zk/datastream/client/stream_client.go

```diff
@@ -357,6 +357,43 @@ LOOP:
 	return err
 }
 
+func (c *StreamClient) ReadBatches(start uint64, end uint64) ([][]*types.FullL2Block, error) {
+	bookmark := types.NewBookmarkProto(start, datastream.BookmarkType_BOOKMARK_TYPE_BATCH)
+
+	marshalled_bm, err := bookmark.Marshal()
+
+	if err != nil {
+		return nil, err
+	}
+
+	c.initiateDownloadBookmark(marshalled_bm)
+
+	batches := make([][]*types.FullL2Block, end-start+1)
+
+	for i := start; i <= end; i++ {
+		for {
+			block, _, batchEnd, _, _, _, err := c.readFullBlockProto()
+			if err != nil {
+				return nil, err
+			}
+
+			if batchEnd != nil && batchEnd.Number == end {
+				break
+			}
+
+			if block != nil {
+				if batches[block.BatchNumber] == nil {
+					batches[block.BatchNumber] = []*types.FullL2Block{block}
+				} else {
+					batches[block.BatchNumber] = append(batches[block.BatchNumber], block)
+				}
+			}
```

Comment:
Instead of checking whether the slice is initialized for certain batch number each time a block is retrieved, we could probably do that only when `batchStart` entry is retrieved from the data stream (slightly more readable IMO). Hopefully there would not be empty batches, because in that case, your approach is better (because it instantiates slice more lazily). WDYT?

```suggestion
			block, batchStart, batchEnd, _, _, _, err := c.readFullBlockProto()
			if err != nil {
				return nil, err
			}

			if batchEnd != nil && batchEnd.Number == end {
				break
			}

			if batchStart != nil {
				batches[block.BatchNumber] = []*types.FullL2Block{}
			}

			if block != nil {
				batches[block.BatchNumber] = append(batches[block.BatchNumber], block)
			}
```